app/jobs/shipit/update_github_last_deployed_ref_job.rb

```diff
@@ -0,0 +1,41 @@
+module Shipit
+  class UpdateGithubLastDeployedRefJob < BackgroundJob
+    queue_as :default
+
+    DEPLOY_PREFIX = 'shipit-deploy'.freeze
+
+    def perform(stack)
+      stack_sha = stack.deploys.where(status: "success").last.until_commit.sha
```

Comment:
You should put that query in a `Stack` method, so that it both have a meaningful name, and is easy to unit test in `stacks_test.rb`.

Also there's already a bunch of methods doing something very close.

There is also a scope for success status: https://github.com/Shopify/shipit-engine/blob/6422110b805c88ee5bd40936ae2b46b06a4e6b0f/app/models/shipit/task.rb#L29

Reply:
Yeah, it's definitely worth extracting. I'll do so now.

Though I figure I should explain why I didn't re-use the existing scopes and methods:

---

I was previously trying to retrieve this information via `stack.last_deployed_commit` (which actually wasn't restrictive enough after talking to Jack, but that's a separate issue.)

That method is just calling `last_completed_deploy`, which itself is just `deploys_and_rollbacks.last_completed`. `last_completed` is just a class method returning the last record of the `completed` scope. That scope includes 'success'.

This worked fine in automated tests. However, I observed some odd behaviour when testing shipit itself locally.

When the Deploy transitioned to Success, it would trigger my method and enqueue the job. However, when the Job later called into the stack (via `last_completed_deploy`), it would return the deploy _prior_ to the one that triggered the job.

But  `stack.deploys` returns the latest deploy correctly.

---

I'm new to Rails, so I wasn't sure if this was due to how scopes work under the hood, or some other mechanism that I wasn't understanding.

I haven't gone so far as looking at the queries themselves to determine exactly what's going on here, but that was a reason I avoided using the existing scope here.

I have a bit more bandwidth now to test around and figure out this behaviour (or discover the flaw in my testing that led me to think this was happening).

Reply:
I did some more research and talked to Jack about the stuff I described above.

It seems that state_machine's `after_transition` executes before the active record commits, and so the job is enqueued before the commit as well.

When running Shipit locally, this can result in the Job itself executing before the Deploy's change to 'success' has been committed, and so the scope filtering to the most recent deploy of 'success' will return the incorrect deploy.

This race condition explains why I was seeing the odd behaviour described above. This is actually an issue anywhere we're using after_transition to enqueue a job that accesses db state.

---

After talking to Jack, we've decided that this is unlikely to occur in production. So I'm going to defer to adding a Stack method that uses the aforementioned scope.

There are a few different overall approaches and workarounds that could be taken, but Jack and I agreed that they should be held off and done on a later change.